### ISSUE TITLE
Puma::HttpParserError when using quotation mark in URI path

### ISSUE THREAD

**Issue by richardboehme** [puma/puma#3396]
**Describe the bug**
We noticed a lot of Puma-related errors in one of our production deployments while being attacked and we were wondering if Puma behaves correctly in this case.

In our case the attacker used a quotation mark (") inside a URL parameter value (e.g. `test="`). This raised the following error:
```
2024-05-31 09:40:57 +0200 HTTP parse error, malformed request ("GET /" - (-)): #<Puma::HttpParserError: Invalid HTTP format, parsing fails. Are you trying to open an SSL connection to a non-SSL Puma?>
```

As far as I understand the URI spec, the quotation mark character isn't a supported character for URI paths, but we still think that Puma should not raise such an error in this case, because other characters that are not mentioned in the spec also do not cause errors. If an error should be raised I think the error message is pretty misleading. 

**Puma config:**

I could reproduce this issue without a custom config.

**To Reproduce**

Run the simple "Hello World"-App (hello.ru):

```
run lambda { |env| [200, {"Content-Type" => "text/plain"}, ["Hello World"]] }
```

Run it with:

```
bundle exec puma --no-config hello.ru
```

Execute the following request against the App:

```
curl -v 'http://localhost:9292/"'
```

**Expected behavior**

Puma does not raise an error and handles this case normally or Puma should raise a more descriptive error. If this is expected behavior, feel free to close the issue.

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04 (in production we do run Alpine 3.17)
 - Puma Version: 6.4.2 (in production we run Puma 5.6.8)

**Comment by dentarg** [comment]
I think you should be encoding that character when used in the URI? What happens in the browser?

**Comment by richardboehme** [comment]
> I think you should be encoding that character when used in the URI? What happens in the browser?

Yeah as I said, I think the quotation mark is not a valid character as defined in the URI spec. Any user facing client (like browsers) would convert it to %22. 

I'm not sure if the behavior of Puma is correct here. However it seems like Puma supports other invalid URIs like "http://localhost:3000/€". In this case it seems like someone in the stack automatically converts the special character into another format. This is how the request path looks like in the env when requesting with the €-sign:
```
{
  ...
  "REQUEST_PATH"=>"/\xE2\x82\xAC",
  ...
}
```

Another example would be curly brackets. Per spec they are not allowed but Puma allows them. For reference, Ruby's `URI.parse` does not allow them.

**Comment by dentarg** [comment]
Interesting if it is inconsistent. Sounds like someone needs to investigate the difference between theses cases then, to explain this.

**Comment by nateberkopec** [comment]
Important to note Puma returns 400 Bad Request in this case. This is correct. If the client provides a malformed URI, we should tell the client and say this request is malformed.

I also think the error isn't misleading. Puma failed to parse the request because it's not valid. "parse error, malformed request" is perfectly descriptive.

I _wish_ we could say exactly what went wrong when parsing, but that's a story for another day perhaps. It would require us to be more sophisticated when raising out of the parser/native extension, something that would require a big overhaul.

It does look like there's probably several inputs re: URI that we _should_ reject but don't. I'm all for fixing some/all of those.

**Comment by richardboehme** [comment]
> Important to note Puma returns 400 Bad Request in this case. This is correct. If the client provides a malformed URI, we should tell the client and say this request is malformed.

I totally agree with this. Puma should either allow all characters and convert them or implement the HTTP spec and be strict about all disallowed characters and I get behind why the latter is the better solution for a library like Puma. 

> I also think the error isn't misleading. Puma failed to parse the request because it's not valid. "parse error, malformed request" is perfectly descriptive.

Well I think the log message "parse error, malformed request" is perfectly descriptive. It may not be specific enough though. It took me quite some time to figure out what the problem was, especially because the full URI was not printed to the log. I understand that logging more specific messages is hard and probably not a big priority right now. Maybe it would help if the exception contained more information about the request, like the full URI?

Also I feel like the "Are you trying to open an SSL connection to a non-SSL Puma?" was quite misleading for me, but I get that this is probably the issue that happens more often, so it makes sense to provide this hint for the most common cases.

**Comment by greytape** [comment]
Did a little bit of investigation here. The issue seems to stem from this HTTP Parser code, which specifically flags `"` as unsafe (along with `<`, `>` and a few others). 
https://github.com/puma/puma/blob/5a8e48570b8c3c64efce75c62c06915f6fdae3f9/ext/puma_http11/http11_parser_common.rl#L14-L16
Any other random characters are just % encoded, but these "unsafe" ones will throw an error.

**Comment by dentarg** [comment]
Some observations:

I ran Puma like this `echo 'app { [200, {}, ["OK"]] }' | ruby -Ilib bin/puma --config /dev/stdin --log-requests`

* for `curl -v 'http://localhost:9292/"'` 
  * no request was logged but the error
  * `2024-10-04 22:31:30 +0200 HTTP parse error, malformed request ("GET " - (-)): #<Puma::HttpParserError: Invalid HTTP format, parsing fails. Are you trying to open an SSL connection to a non-SSL Puma?>` was logged, as reported by this issue
* for `curl -v 'http://localhost:9292/€'`
  * `127.0.0.1 - - [04/Oct/2024:22:33:00 +0200] "GET /%e2%82%ac HTTP/1.1" 200 - 0.0004` was logged
* for `curl -g -v 'http://localhost:9292/{}'`
  * `127.0.0.1 - - [04/Oct/2024:22:33:32 +0200] "GET /{} HTTP/1.1" 200 - 0.0001` was logged

I also tried `curl -v 'http://www.google.com/"'`, Google returns 404 but the response body does include `<p>The requested URL <code>/%22</code> was not found on this server.`

For `curl -v 'http://www.google.com/€'` the response included `<p>The requested URL <code>/%E2%82%AC</code> was not found on this server.`

For `curl -g -v 'http://www.google.com/{}'` the response included `<p>The requested URL <code>/%7B%7D</code> was not found on this server.`
